### PR TITLE
Persistent RNN functions in cuDNN v6

### PR DESCRIPTION
**Pull request by aonotas**
This PR is related cupy PR ( https://github.com/cupy/cupy/pull/549 ).

Note that only CA (Compute Capability) > 6.0 supports Persistent RNN.
https://developer.nvidia.com/cuda-gpus

I have checked this PR in cuDNN v6 and v7.
- algo=standard (v6, v7)
- algo=static (v6, v7)
- algo=dynamic (v6, v7)

### CODE REVIEW

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by aonotas** [comment]
I think that travis error is not related this PR.

File: chainer/functions/connection/n_step_gru.py [hunk 1]
```diff
@@ -25,20 +25,23 @@
...
+    def __init__(self, n_layers, states, rnn_algo, **kwargs):
         n_step_rnn.BaseNStepRNN.__init__(
-            self, n_layers, states, rnn_dir='bi', rnn_mode='gru', **kwargs)
+            self, n_layers, states, rnn_dir='bi', rnn_mode='gru',
+            rnn_algo=rnn_algo, **kwargs)
 
 
 def n_step_gru(
-        n_layers, dropout_ratio, hx, ws, bs, xs, **kwargs):
+        n_layers, dropout_ratio, hx, ws, bs, xs, rnn_algo='standard',
```
Comment (unnonouno) [inline chainer/functions/connection/n_step_gru.py]: fix docstring

File: chainer/functions/connection/n_step_gru.py [hunk 2]
```diff
@@ -119,11 +122,13 @@ def n_step_gru(
...
 
     return n_step_gru_base(n_layers, dropout_ratio, hx, ws, bs, xs,
-                           use_bi_direction=False, **kwargs)
+                           use_bi_direction=False,
+                           rnn_algo=rnn_algo, **kwargs)
 
 
 def n_step_bigru(
-        n_layers, dropout_ratio, hx, ws, bs, xs, **kwargs):
+        n_layers, dropout_ratio, hx, ws, bs, xs, rnn_algo='standard',
```
Comment (unnonouno) [inline chainer/functions/connection/n_step_gru.py]: ditto

File: chainer/functions/connection/n_step_lstm.py [hunk 3]
```diff
@@ -25,20 +25,23 @@
...
+    def __init__(self, n_layers, states, rnn_algo):
         n_step_rnn.BaseNStepRNN.__init__(self, n_layers, states,
-                                         rnn_dir='bi', rnn_mode='lstm')
+                                         rnn_dir='bi', rnn_mode='lstm',
+                                         rnn_algo=rnn_algo)
 
 
 def n_step_lstm(
-        n_layers, dropout_ratio, hx, cx, ws, bs, xs, **kwargs):
+        n_layers, dropout_ratio, hx, cx, ws, bs, xs, rnn_algo='standard',
```
Comment (unnonouno) [inline chainer/functions/connection/n_step_lstm.py]: ditto

File: chainer/functions/connection/n_step_rnn.py [hunk 4]
```diff
@@ -161,8 +173,20 @@ def __init__(self, n_layers, states, rnn_dir, rnn_mode, **kwargs):
...
                              % (rnn_mode, candidate_list))
+
+        is_enable_cudnn_v6 = isinstance(_rnn_persistent_algo, dict)
+        if is_enable_cudnn_v6 and rnn_algo not in _rnn_persistent_algo:
+            candidate_list = ','.join(_rnn_persistent_algo.keys())
+            raise ValueError('Invalid rnn_algo: "%s". Please select from [%s]'
+                             % (rnn_algo, candidate_list))
+            _rnn_algo = _rnn_persistent_algo[rnn_algo]
+        else:
+            _rnn_algo = None
```
Comment (unnonouno) [inline chainer/functions/connection/n_step_rnn.py]: With this code, when a user uses older cuDNN, rnn_algo argument is ignore silently, isn't it? It is better to notify this argument is unavailable in such situation. Check if a user specify 'static' or 'dynamic' and when he/she uses older cuDNN raise an error.

File: chainer/functions/connection/n_step_rnn.py [hunk 5]
```diff
@@ -303,8 +327,19 @@ def forward(self, inputs):
...
         rnn_desc = cudnn.create_rnn_descriptor(
             n_units, self.n_layers, self.states._desc,
-            libcudnn.CUDNN_LINEAR_INPUT, self.rnn_dir,
-            self.rnn_mode, libcudnn.CUDNN_DATA_FLOAT)
+            libcudnn.CUDNN_LINEAR_INPUT, self.rnn_dir, self.rnn_mode,
+            libcudnn.CUDNN_DATA_FLOAT, algo=self.rnn_algo)
+
+        if self.rnn_algo == libcudnn.CUDNN_RNN_ALGO_PERSIST_DYNAMIC:
+            batchsize = len(x_list[0])
+            global _prev_batchsize
```
Comment (unnonouno) [inline chainer/functions/connection/n_step_rnn.py]: Is it thread safe?

File: chainer/functions/connection/n_step_rnn.py [hunk 6]
```diff
@@ -144,10 +144,22 @@ def get_random_state():
         libcudnn.CUDNN_LSTM: True,
     }
 
+_rnn_persistent_algo = None
+if cuda.cudnn_enabled and _cudnn_version >= 6000:
+    _rnn_persistent_algo = {
+        'standard': libcudnn.CUDNN_RNN_ALGO_STANDARD,
+        'static': libcudnn.CUDNN_RNN_ALGO_PERSIST_STATIC,
+        'dynamic': libcudnn.CUDNN_RNN_ALGO_PERSIST_DYNAMIC,
+    }
```
Comment (unnonouno) [inline chainer/functions/connection/n_step_rnn.py]: How a user select 'static' or 'dynamic'? What's the difference?

Comment (aonotas) [reply chainer/functions/connection/n_step_rnn.py]: 'static' only supports small network, but 'dynamic' supports both small and large networks.
'dynamic' need to call `create_rnn_persistent_rnn_plan `.  This function needs computation cost. So we need to reuse it.

File: chainer/functions/connection/n_step_rnn.py (hunk 5 above)
Comment (aonotas) [reply chainer/functions/connection/n_step_rnn.py]: Do you have any idea to cache `_rnn_plan ` when a batchsize is same?

**Comment by unnonouno** [comment]
As we discussed offline, we found it is not enough fast with the current implementation. I think our implementation has another bottle neck especially when mini-batch size is small. Before merging this PR, we need to improve the current implementation.
BTW, i refactored RNNs. Please check the latest one.

**Comment by aonotas** [comment]
OK. Thank you for refactoring!!
After papr submission (deadline is 2/23 in Japan), I will fix code.

**Comment by aonotas** [comment]
It seems that when we use `rnn_algo` = 'static' or 'dynamic', output does not correspond to `standard`.
https://gist.github.com/aonotas/be96e909e2382013c4c0df5ebc844525

I will support `rnn_algo` = 'standard' only.

Branch
Chainer : https://github.com/aonotas/chainer/tree/support_persistent_rnn_cudnn
Cupy : https://github.com/aonotas/cupy/tree/support_persistent_rnn_cudnn

I close this PR.

**Pull request closed without merging**